### PR TITLE
POC / Proposal: Package namespaces

### PR DESCRIPTION
**Pull request by davidmdm**
# Proposal / POC

## Overview

It would be great to be able to create rules that do not collide with the org package namespace and that would directly be mapped to a specific meta condition.

Although it is possible today to write rules that apply to specific projects/branches using `data.meta` and toggling rule enablement, everything lives in one giant namespace: `org`. 

This PR allows for new set of valid package names:
- org
- project[expression] // where expression can be the project UUID or a shell-expression of a project slug
- branch[expression] // where expression is a shell-expression of a branch name

### Some implications:
- all rule name evaluations now get prefixed by their matched package. For example rule `custom` in package `org` will report as a rule violation of `org.custom` and rule `custom` in package `branch.main` will report as a rule violation of `branch.main.custom` and so on.
- Rule name collisions cross packages supplement one another. One does NOT override an org level rule with a project or branch based rule of the same name.

### Current implementation tradeoffs
- The policy-agent in its current form has become overly specific to `Config Policies`. A serious backwards compatibility breaking refactor would need to take place for the agent to support other domain specific policies that do not follow `org`, `project`, `branch` semantics.
- Pattern matching uses path globs for simpler expressions and less compute overheard. However it gives less power than a regexp would. A simple example of this would be: `package branch.["^main"]` to apply to any branch that is not `main`. However these kind of negative groupings can be easily expressed at the org level with rule guards. Regexp vs Glob support is still an undecided thing at this point. The POC goes with Glob support.

## Examples

With this we could write a policy for a project without having to guard every rule and that can have symbols that conflict with org ones:

```rego
package project["b02c619e-5958-11ed-a1a4-ff0a876360a6"]

enable_rule["conflicts_with_org_but_thats_ok"]
```

or for a given set of slugs:

```rego
package project["prefix-*"]
...
```

or for a branch expression:
```rego
package branch["feature-*"]
...
```

# Proposal V2

There's a secondary proposal (not implemented) for an alternative way of matching package to metadata.
This proposal has a more generic structure and is more expressive and can map packages to multiple metadata conditions but can a little more confusing from the user's perspective.

Simply put we can encode all of the metadata conditions as series of parameters like url query parameters in some predictable package structure, using the fact that we can have arbitrary strings as package name segments.

Therefore we could introduce a secondary valid package name like `package org.filters[expr]` or `package filters[expr]`
and let the expression be a set of meta conditions.

Thus we could have:

```rego
package org.filters["branch=main&project_slug=gh/team-prefix-*"]
```

And create packages based on multiple meta conditions.

## Tradeoffs
- More flexible and powerful
- Not as predictable: multiple packages could express the same conditions but have different names
- PRO & CON: Packages names are not semantic to the policy domain

### CODE REVIEW

**Comment by eddiewebb** [comment]
Thank you for sharing @davidmdm !

Firstly, I like this proposal, I think scoping of packages is a great thing to figure out.  I don't see the current model technically challenge but a mental hurdle.  Explicit packaging that indicates scoping would help overcome that I think.

As for the 2 proposals, **the concern I have with #1,** is that they are not exclusive domains, where as a 'package' generally should be right? (how can a cat be in 2 boxes at once!?)

Anyway what i mean is while project is a logical subset of org (org.project), I am not convinced branch is, because the branch can only exist inside a project (org.????.branch).

So for proposal 1 I think to be clean means we would maybe want a global vs project package (org.branches.main, org.project.branches.main)  and I am  note sure that keeps the simplicity you endorsed.  


Contrary, **I agree very much with your concerns on #2**, while it solves mine I suppose

Dumb question from the guy that doesn't have to implement, but couldn't we just assert canonical naming based on the query string?

i.e. what if we asserted that however you write it, we do some ordering and defaulting of those values (PS i do think we should limit how many filters we support vs expecting to be done as part of policy itself, I see limited value in solving for all possible filters in package names - tag? nothing else.) 

| expression | our value |
------------------------
| org.filters["branch=main"] |  {"projects":"*","branches":"main"} |
| org.filters["project_slug=gh/team-prefix-*"] |  {"projects":"gh/team-prefix-*","branches":"*"} |
| org.filters["branch=main&project_slug=gh/team-prefix-*"] |  {"projects":"gh/team-prefix-*","branches":"main"} |
| org.filters["project_slug=gh/team-prefix-*&branch=main"] |  {"projects":"gh/team-prefix-*","branches":"main"} |


This solves my concern by being explicit about where branch rules are being applied.

**SUmmary:**

I probably lean towards #2 if we can get canonical, but 
I think the branch filtering makes me uncomfortable given overlap across projects, I would rather see that defined within policy, not only for overlap reason above but also default branch and branch flow (i.e main, dev, current, etc) will vary by team within an org.
